### PR TITLE
Load bundled GraalJS, give priority to GraalVM if present.

### PR DESCRIPTION
**Pull request by sdedic**
# Description
This PR fixes several code loading bugs that manifested in Maven/Gradle proxy support **not working** if NB IDE (or vscode extension) runs on a recent GraalVM JDKs. The IDE was not able to load JS interpreter and failed to process PAC to discover the proxy.

The reason was that because subtle changes in implementation over the time, GraalVM JDK implementation uses `ServiceLoader` to discover even parts of itself - but intentionally restricts classloaders to GraalVM JDK modules. So once Polyglot SDK class loads from the GraalVM JDK, the polyglot implementation may not access NB-distributed GraalJS implementation, although it is formally registered as a polyglot language. Some time ago, GraalVM stopped to ship `js` with the JDK, and must be installed. Until that change, it was OK (for the purpose of JS processing in NB) that the Polyglot engine + JS implementation loaded from the JDK. 

After that change, it isn't OK: as classloaders in NB delegate first to the parent (ultimately to the bootstrap classloader), it would load classes from  Polyglot API and parts loaded by name through `ServiceLoader` from the JDK instead of NB, and bypass shipped Javascript, although the JDK does not contain `js` implementation at all.

# Module system tweaks
There are two tweaks in this PR:
1. `eager` modules can now declare `OpenIDE-Module-Package-Dependency:` and `OpenIDE-Module-Java-Dependency` in their manifests. It was never documented that for eager this should not work for `eager` modules, and other unsatisfied dependency caused (correctly) the eager module to be silently ignored and not enabled. But Package dependency was not check at first, only at very late stages - and the failure caused errors in the log and during startup.
2. For the implementation, I've needed an instance of `ProxyClassLoader`, that delegates the `shouldDelegateResource` to either a subclass or some external code. the PR adds a second constructor for `ProxyClassLoader` - it is used by reflection from `libs.graalsdk`.
3. For `autoload` modules that provide tokens, the same check as for `eager` modules was added: this allows early detection of autoloads with unsatisfied java/package dependencies such as `graalvm.sdk` and 'javafx.linux`. 

We probably need a proper API for such custom loaders - I take this as an action item for NB19. The use-case to construct a ClassLoader with flexible parent-first policy that sees some dependency closure seems valid, and the optimizations inside bootstrap are too important and hard to reproduce.

# The fix itself
The fix itself is in `libs.graalsdk`, where `GraalEngine` wrappers around polyglot languages are created.  To work with polyglot API, the module creates a custom classloader that includes **all its dependents**, so a `ServiceLoader` is able to enumerate the languages and the polyglot internal implementation itself. 

If this was not done, `ServiceLoader` would use `SystemClassLoader` that contains a special logic: in order to avoid possible errors when multiple instances of the same-named class are exported by different modules, it attempts to identify `owning module` of the Class being linked (or requesting the load) and searches **in its owning module's closure** upwards and including the JDK.
In this case, this means that it will start search from GraalSDK module and its dep closure. But GraalSDK 3rd party library attempts to search for an implementation that actually resides in `libs.truffle` module that **depends on graalsdk**, so it is NOT in `graalsdk`'s dependency closure. Under OpenJDK, this fails, and full breadth search through all modules is done by `SystemClassLoader`.
Under GraalVM JDK, the Truffle implementation is found in the JDK and loads from there ... and then fails to reach to NB-distributed JS implementation, if GraalVM JDK does not one installed in it.

With the custom ClassLoader, the `ServiceLoader` properly locates all parts shipped with NetBeans even in presence of GraalVM JDK. This custom `ClassLoader` is created from the `ProxyClassLoader` using reflection, to avoid immature API change. I take an action item to develop a proper API in early NB19 timeframe.

To bridge **real GraalVM implementation**, a new module, `libs.graalsdk.system` is introduced. This module reaches to JDK implementation of Polyglot, Truffle and whatever languages and makes bridges the same way as `libs.graalsdk`.

I've adjusted the Scripting implementation so that **if GraalVM implementation is present**, the GraalVM JS takes precedence in by-extension, by-name and by-mime lookups as well as when enumerating all `ScriptEngineFactories`. This is also covered by tests.

# Test infrastructure fix
Our `NbModuleSuite` is based on ancient JUnit3, and has no support for `assumeXXX` methods that classify tests as not applicable to whatever runtime conditions of the testing environment. If one uses `assumeTrue` in a `NbTestCase`, the result is  an error (not failure). I've adjusted `NbTestCase` to handle failed assumption as a success. For that I've needed to publish `org.junit.internal', as `AssumptionViolatedException` derives from an exception from that internal package,  and if the superclasss is not visible, the exception is not actually usable in source code -- the compiler fails on an unreachable superclass.

# Test adjustments
All the `libs.graalsdk`, `api.scripting` tests run from the (application) classpath. That was fine when no parent-classloader-ignoring magic was in place, but now would not work. I've converted them all into `NbModuleSuite` tests that run NB module container and load tests from within it.

# The proper fix (plan)
@jtulach has suggested that the search among dependents is not a new idea, that Eclipse container solves the same problem in similar scenarios. So the suggestion was to adopt its approach, and be explicit about the mutual agreement to load from dependents subtree, and being an depenedent. See the ["buddy" attribute](https://wiki.eclipse.org/Context_Class_Loader_Enhancements#Buddy_Class_Loading).

### CODE REVIEW

File: harness/nbjunit/src/org/netbeans/junit/NbTestCase.java
Comment (mbien) [inline harness/nbjunit/src/org/netbeans/junit/NbTestCase.java:480]: awesome. This will make the test case show up as green though, not skipped, right?

I actually wanted to implement this myself but got distracted before I figured out how to communicate this to junit properly.

Comment (sdedic) [reply harness/nbjunit/src/org/netbeans/junit/NbTestCase.java:480]: Sadly - I didn't come with a way how to mark it skipped, here in JUnit3-style code. The actual `TestResutl` that came around at runtime only allowed to mark the test passed/failed/error. I am not particularly happy with marking skipped as passing but better than failure. 

An option would be to rewrite all those `assume` statements to if-returns.

Please suggest / elaborate.

**Comment by sdedic** [comment]
Note to self: merge fixing commits into their original feature commits.

**Comment by neilcsmith-net** [comment]
Added the do-not-merge to track status of this.  Please remove the label as and when the review is complete, and the PR branch is squashed and force pushed to the state you want it merging in.

Are there any risks to the module system tweaks?  We'll be looking to close off delivery on Wednesday for 18-rc2 - might be good to get this and tested sooner rather than later.

**Comment by sdedic** [comment]
> Are there any risks to the module system tweaks? We'll be looking to close off delivery on Wednesday for 18-rc2 - might be good to get this and tested sooner rather than later.

For the classloading no - the new code path is only triggered from within GraalSDK module for its operation and for a specific private ClassLoader instance. No other ClassLoader should be affected.

As for the `eager` and `autoload` tweaks, there's a slight chance that it changes behaviour, especially the first start - varying on different systems. But it seems easy to test:
- there should not be any complaint that a module cannot be enabled in the startup log (except possibly javafx)
- if JDK runtime supports javafx, exactly one module (for the runtime platform) should load
- it's best to test **without** ergonomic cluster with all standalone modules enabled

The fix itself can be easily tested on a network where PAC is present (http://wpad/wpad.dat exists and produces content):
- PAC is processed on standard JDK
- PAC is processed on GraalVM JDK **without** javascript installed
- PAC is processed on GraalVM JDK **with** javascript installed
(these situations are covered by unit tests for `libs.graalsdk` and `libs.graalsdk.system` modules

**Comment by neilcsmith-net** [comment]
Thanks for the info @sdedic Do you still intend this for NB18?  Will need the review complete and PR updated for merge this week to target rc3, which will hopefully be the last release candidate for 18.

cc/ @MartinBalin

**Review by MartinBalin — APPROVED** [review]

**Comment by MartinBalin** [comment]
Yes, it will be great to have it in NB18 it was tested, is working....

**Review by dbalek — APPROVED** [review]
Seems to work fine for me.

**Comment by mbien** [comment]
as reminder this needs squashing to make it ready -> https://github.com/apache/netbeans/pull/5842#issuecomment-1520483532

**Comment by sdedic** [comment]
Rearranged commits.

**Comment by neilcsmith-net** [comment]
Thanks @sdedic Assuming ready to merge once tests complete now?

**Comment by neilcsmith-net** [comment]
@sdedic it looks like not all the relevant tests were enabled on this one?  It looks like LSP and Graal tests were skipped but have been failing in delivery since - see https://github.com/apache/netbeans/actions/runs/4873351185 (ignore PHP one)

cc/ @MartinBalin - will need to consider this from release team point of view, and if it definitely caused by this whether we need to revert for rc3.

**Comment by sdedic** [comment]
whoops. Let me check.

**Comment by neilcsmith-net** [comment]
@sdedic thanks!  As just mentioned in email on dev@ there also seem to be some random failures in other tests that were stable beforehand, looking at the history of full testing on merges to delivery before and after this was merged.

**Comment by sdedic** [comment]
FYI - I seem to have found the cause; there seem to be quite a change in behaviour between Graal 22 and Graal 20 in terms of classloading. I need to give the patch one more full round of tests (Graal 20.3, Graal 21.1, Graal 22.3, OpenJDK) locally, but should be ready for review tomorrow.

**Pull request merged**
